### PR TITLE
Allow multiple directories

### PR DESCRIPTION
**Pull request by spudowiar**
Work in progress and currently multiple `layouts_dir`s and `includes_dir`s has been implemented.

This is necessary for #4510 and is more generic than the current implementation in #4595 in that it allows you to list multiple `layout_dirs`, whereas #4595 manually deals with a single case.

### CODE REVIEW

**Comment by parkr** [comment]
@spudowiar Please ping me directly when this is finished. Thanks!

**Comment by spudowiar** [comment]
Lovely! Hit by #4250!

**Comment by spudowiar** [comment]
@benbalter @parkr Other than `_layouts` and `_includes`, which other directories need to be implemented?

**Comment by benbalter** [comment]
> Other than _layouts and _includes, which other directories need to be implemented?

I believe just those two. For https://github.com/jekyll/jekyll/pull/4595, [sass has it's own mechanism](https://github.com/twbs/bootstrap-sass/blob/master/lib/bootstrap-sass.rb#L62) which we can use, and `_assets` or other static files will need to be added to static files separately.

**Comment by spudowiar** [comment]
I was looking at SASS but it's mechanism only works when `!site.safe` so it will still need implementing.

**Comment by benbalter** [comment]
> I was looking at `jekyll-sass-converter` but its mechanism only works when `!site.safe` so it will still need implementing.

It should work in safe mode (I use it for my site on Pages). I believe *some* configuration options are disabled via YAML, but we can override them via Ruby.

**Comment by spudowiar** [comment]
In `jekyll-sass-converter` the only way to have multiple SASS directories is disabled when `site.safe` (see https://github.com/spudowiar/jekyll-sass-converter/blob/master/lib/jekyll/converters/scss.rb#L81)

Overriding it in Ruby is not a good idea as I intend for the same functionality to be achieved using `sass_dir` as an array and we would have to sanitise the path anyway so why not just do it there.

**Comment by spudowiar** [comment]
@benbalter @parkr ETA on this being merged?

**Comment by envygeeks** [comment]
hulk: smash +security +performance

**Comment by envygeeks** [comment]
I have done what you requested, good luck. /cc @envygeeks

**Comment by envygeeks** [comment]
Other then my bot needing some consistency fixes I see no major problems.  I'm :+1:

File: test/source/_layouts2/default2.html [hunk 1]
```diff
@@ -0,0 +1,25 @@
+<!DOCTYPE html>
+<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en" lang="en-us">
+<head>
+   <meta http-equiv="content-type" content="text/html; charset=utf-8" />
+   <title>{{ page.title }}</title>
+   <meta name="author" content="<%= @page.author %>" />
```
Comment (pathawks) [inline test/source/_layouts2/default2.html]: What's with the ERB tag?

Comment (spudowiar) [inline test/source/_layouts2/default2.html]: It's from the existing `default.html`, I just duplicated it

Comment (parkr) [inline test/source/_layouts2/default2.html]: Oh weird, it shouldn't have any ERB?

File: lib/jekyll/readers/layout_reader.rb [hunk 2]
```diff
@@ -7,23 +7,30 @@ def initialize(site)
...
+      end
     end
 
     private
 
     def layout_entries
-      entries = []
-      within(layout_directory) do
-        entries = EntryFilter.new(site).filter(Dir['**/*.*'])
+      entries = {}
```
Comment (parkr) [inline lib/jekyll/readers/layout_reader.rb]: I suppose it is a bit less efficient, but using `reduce` or `inject` would be a bit more elegant, e.g.

```ruby
def layout_entries
  layout_directories.inject({}) do |layouts, directory|
    within(directory) do
      layouts[directory] = \
        EntryFilter.new(site).filter(Utils.safe_glob("*", ["**", "*.*"]))
    end 
    layouts
  end
end
```

File: lib/jekyll/readers/layout_reader.rb [hunk 3]
```diff
@@ -7,23 +7,30 @@ def initialize(site)
...
+        end
       end
 
       @layouts
     end
 
-    def layout_directory
-      @layout_directory ||= (layout_directory_in_cwd || layout_directory_inside_source)
+    def layout_directories
+      @layout_directories ||= Array(site.config['layouts_dir']).map do |directory|
```
Comment (parkr) [inline lib/jekyll/readers/layout_reader.rb]: I am not convinced this should be configurable as an array. I think it's useful for themes, but the `Site` can ask the `Theme` for its directory. I'd rather see this as:

```ruby
def layout_directories
  @layout_directories ||= Array(site.config['layout_dir']) + Array(theme.layout_dir)
end
```

Before #4595, just the first piece will do.

File: lib/jekyll/tags/include.rb [hunk 4]
```diff
@@ -104,14 +104,32 @@ def tag_includes_dir(context)
 
       def render(context)
         site = context.registers[:site]
-        @includes_dir = tag_includes_dir(context)
-        dir = resolved_includes_dir(context)
+        @includes_dir = Array(tag_includes_dir(context))
+
+        includes_dir = @includes_dir.map do |dir|
+          resolved_includes_dir(context, dir)
+        end
```
Comment (parkr) [inline lib/jekyll/tags/include.rb]: We already have `include_relative`, why do we need this?

**Comment by parkr** [comment]
@spudowiar Ok, I have taken a look at this. So far, I am not convinced that this is required for gem-based plugins. The problem you're trying to solve (if I'm understanding correctly) is what @benbalter is trying to solve in his PR as well: being able to bring in resources from external sources. In terms of security and ease-of-use, I think his route will serve the community better. If a firm (such as 18F, GitHub, or even Jekyll) wishes to ship layouts or includes, they will be able to do so in a versioned gem. It solves the same problem very similarly but with some added perks.

Maybe @jekyll/stability and @jekyll/ecosystem have some thoughts about this?

File: lib/jekyll/readers/layout_reader.rb (hunk 2 above)
Comment (envygeeks) [inline lib/jekyll/readers/layout_reader.rb]: Use each with object, it's faster than inject.

**Comment by spudowiar** [comment]
There are many uses for this as a configuration option for where something is not complex enough to bundle as a theme

**Comment by pathawks** [comment]
> There are many uses for this as a configuration option for where something is not complex enough to bundle as a theme

What are some use cases you have in mind?

**Comment by spudowiar** [comment]
>What are some use cases you have in mind?

If you have, for example, two sites which share layouts, it's far easier to simply have a `_shared` directory that is a symlink or a Git submodule and adjust `layouts_dir` in `config.yml` than create a Gem containing these files and configuring that (also, what if you already have a theme set?)

**Comment by pathawks** [comment]
It seems like Git submodules is exactly tge type of problem that Themes are aiming to solve.

**Comment by spudowiar** [comment]
Themes are a great idea, someone creates a theme to be used by lots of people, lots of people use it, no version fragmentation, etc.

They are far too heavy for simple projects though. The problem with Git submodules is that people don't keep them up to date, they modify it and there are merge conflicts and messes etc.

But if you have a simple project, you don't need all this overhead, you can just have a simple shared folder, far simpler. It's a lot easier for most users, who won't know or want to know how to package a Gem and allows for quicker changes. (For a Gem, `modify file -> package gem -> install gem -> rebuild site`, for a shared folder it's `modify file -> rebuild site (optional, --watch also works)`

**Comment by pathawks** [comment]
But a "Gem" can also just be a local folder; its not something you have to package & install if you’re only using it yourself.

**Comment by spudowiar** [comment]
>But a "Gem" can also just be a local folder; its not something you have to package & install if you’re only using it yourself.

I don't understand, this uses RubyGems, how would you use a local folder? But still, this is a far simpler method

**Comment by pathawks** [comment]
In your Gemfile:

```Gemfile
gem 'MyJekyllTheme', :path => '../MyJekyllTheme/'
```

**Comment by spudowiar** [comment]
But still, you have to create a `Gemfile` or `.gemspec` for this folder

**Comment by parkr** [comment]
Ok, after talking for a while with @benbalter, I'm happy with the decision to move forward with Gem-based themes *only*. The advantages to gem-based themes are significant:

1. Less prone to errors when sharing either your site or your theme ( 🌟 🌟 🌟 🌟 🌟 )
2. We will ship a command, `jekyll new-theme path_to_theme`, which makes the creation of the files and directories a one-step process. It will include instructions as well as a scaffold.
3. Private, ad-hoc themes are still supported through github.com (private repo) or local (`:path` in Gemfile) distribution.
4. A little added overhead for the theme creator makes it _much_ simpler for users. Versioning (including, crucially, the ability to roll back to a previous version) is paramount here.

This decision took me a couple weeks of thought which is much longer than I usually take for these sorts of things. I read all the comments above and talked it out at length with Ben, arguing that it should be configurable. Ultimately, he convinced me that the need for a wonderful user experience for the consumer of a theme (not the producer) over many months and years is what is important. Cloning a repository to some random location on your machine and pointing your config there is not a sustainable way to solve the problem of shared layouts, includes, and styles.

As an implementation detail, we will have to setup load paths for layouts & includes anyway, so a plugin could easily inject more paths if it so chose.

With that, I'm going to close this PR. @spudowiar, feel free to add support you added here for multiple directories to #4595, leaving out the configuration via `_config.yml`.

Thanks for the great discussion!

**Comment by spudowiar** [comment]
@parkr Provided the internal implementation remains stable, I'm perfectly happy with this decision

**Comment by parkr** [comment]
> @parkr Provided the internal implementation remains stable, I'm perfectly happy with this decision

@spudowiar It might take a few iterations to make it fully stable but we'll try to make migration to the new API really obvious and painless for plugin authors.

**Comment by spudowiar** [comment]
@parkr I just want to be able to expose this functionality through a plugin that won't break randomly

**Comment by parkr** [comment]
@spudowiar Absolutely. We try to do that but sometimes bugs are introduced. Did you join any of our Affinity Teams? https://teams.jekyllrb.com/

**Comment by spudowiar** [comment]
@parkr Yeah, I'm in build and ecosystem

**Comment by parkr** [comment]
@spudowiar Great! Keep an eye on those notifications. We'll be sure to notify if there is anything drastic happening – I don't suspect so but I'll use those teams as a means of ensuring this.

**Comment by spudowiar** [comment]
@parkr Don't worry, been trying to help out on all of the issues and pull requests that come through those Affinity Teams as much as I can :)

**Pull request closed without merging**